ext/bg/js/dictionary.js

```diff
@@ -144,6 +172,118 @@ function dictTermsGroup(definitions, dictionaries) {
+                    score: Number.MIN_SAFE_INTEGER,
+                    expression: new Set(),
+                    reading: new Set(),
+                    expressions: new Map(),
+                    source: definition.source,
+                    dictionary: definition.dictionary,
+                    definitions: []
+                };
+            }
+            const score = Math.max(definitionsBySequence[definition.sequence].score, definition.score);
+            definitionsBySequence[definition.sequence].score = score;
+        } else {
+            definitionsBySequence['-1'].push(definition);
+        }
+    }
+
+    return definitionsBySequence;
+}
+
+function dictTermsMergeByGloss(result, definitions, appendTo, mergedIndices) {
+    const definitionsByGloss = appendTo || {};
+    for (const [index, definition] of definitions.entries()) {
+        if (appendTo) {
+            let match = false;
+            for (const expression of result.expressions.keys()) {
+                if (definition.expression === expression) {
+                    for (const reading of result.expressions.get(expression).keys()) {
+                        if (definition.reading === reading) {
+                            match = true;
+                            break;
+                        }
+                    }
+                }
+                if (match) {
+                    break;
+                }
+            }
+
+            if (!match) {
+                continue;
+            } else if (mergedIndices) {
+                mergedIndices.add(index);
+            }
+        }
+
+        const gloss = JSON.stringify(definition.glossary.concat(definition.dictionary));
+        if (!definitionsByGloss[gloss]) {
+            definitionsByGloss[gloss] = {
+                expression: new Set(),
+                reading: new Set(),
+                definitionTags: new Set(),
+                glossary: definition.glossary,
+                source: result.source,
+                reasons: [],
+                score: definition.score,
+                id: definition.id,
+                dictionary: definition.dictionary
+            };
+        }
+
+        definitionsByGloss[gloss].expression.add(definition.expression);
+        definitionsByGloss[gloss].reading.add(definition.reading);
+
+        result.expression.add(definition.expression);
+        result.reading.add(definition.reading);
+
+        // result->expressions[ Expression1[ Reading1[ Tag1, Tag2 ] ], Expression2, ... ]
+        if (!result.expressions.has(definition.expression)) {
+            result.expressions.set(definition.expression, new Map());
+        }
+        if (!result.expressions.get(definition.expression).has(definition.reading)) {
+            result.expressions.get(definition.expression).set(definition.reading, new Set());
+        }
+
+        for (const tag of definition.definitionTags) {
+            if (typeof tag === 'string') {
```

Comment:
Why again is it that `definitionTags` has tags that are either strings or objects? Is there a reason why they aren't objects always?

Reply:
`dictTermsMergeByGloss` is used for definitions from `Database.findTermsBySequence` first (tags aren't expanded) and a mix of definitions from `Translator.findTerms` (tags expanded) and `Database.findTermsExact` (not expanded) after that. The tag metadata isn't needed in `dictTermsMergeByGloss` currently, so I didn't think it was necessary to expand the tags in the rest of the definitions at that point yet.

Reply:
I think tags should be expanded from the start for everything for clarity sake. It's much harder to reason about code when the type is not explicit. Of course, in languages like JavaScript, explicit type declaration is not an option (unless you use Typescript or something), so we have to rely on convention. By making tags be the same thing everywhere it's a lot easier to write bug-free code. Using `typeof` should be reserved only for checking if something that could have a false-like value is `undefined`.